Move off the release train to a simpler release model (#17)
NullVoxPopuli/rfcs: text/0000-move-off-the-release-train-to-simpler-release-model.md

release-date: # In format YYYY-MM-DDT00:00:00.000Z
release-versions:
teams: # delete teams that aren't relevant
  - cli
  - framework
prs:
  accepted: # Fill this in with the URL for the Proposal RFC PR
project-link:
suite:
---

# Move off the release train to a simpler release model

## Summary

Ember's "release train" today is really two things bundled together:

1. A six-week release cadence. A new minor ships on a steady, predictable clock.
2. A multi-branch promotion pipeline. Code gets promoted from the default branch through dedicated `beta` and `release` branches (plus LTS branches), shepherded by a small number of folks who manage each release.

This RFC keeps (1) and replaces (2). The six-week cadence is the part that works, so it stays exactly as it is. What goes away is the extra release branches and the bespoke, per-cycle process that goes with them.

Concretely:

- Just `main`. No `beta` or `release` branches.
- Stable releases are cut from `main` by [`release-plan`][release-plan] on the same six-week schedule. The `npm publish` is gated behind a protected [GitHub deployment environment][gh-environments], so a maintainer approves before anything publishes.

> - We drop the `beta` channel. A new `@alpha` prerelease is published nightly from `main` (the way Embroider and Glint _used to_ publish theirs), and `ember-source@beta` users either move to `@alpha` or drop the scenario. Canary is unchanged: consuming `main` from git works exactly like it does today.

Reviewer: ```suggestion
- We drop the `beta` channel. `@alpha` prerelease is nearly the same as before, but published nightly from `main`, and `ember-source@beta` users either move to `@alpha` or drop the scenario. 
```

SemVer, the six-week cadence, the deprecation policy, LTS, and the major-version process from [RFC #0830][rfc-830] are all unchanged. The only things that change are the branch structure and the publishing mechanics.

[release-plan]: https://github.com/embroider-build/release-plan
[gh-environments]: https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment
[rfc-830]: https://github.com/emberjs/rfcs/blob/master/text/0830-evolving-embers-major-version-process.md

## Motivation

> The expensive part of the release train isn't the cadence. It's all the machinery around it, plus the time it takes maintainers with limited availability to keep it running.

Reviewer: rephrase this, no one brought in to question the cadence

### Releases depend on too few people

> We don't actually have a formal release-manager rotation. In practice the release comes down to essentially one person per project: Katie for `ember.js`, Chris for `ember-cli`. That's a real bus-factor and burnout risk, because when that person isn't around, the release slips.

Reviewer: ```suggestion
We don't actually have a formal release-manager rotation. In practice the release comes down to essentially one person per project: @katiegengler for `ember.js`, @mansona for `ember-cli`. This is a bus-factor and burnout risk -- which other teams have already faced. When that person isn't around (due to life, unplanned things, etc (this alone is not a big deal)), the release slips.
```

> The biggest win of getting off the train is that releasing stops being specialized knowledge. With [`release-plan`][release-plan], _any_ maintainer can trivially cut a release: they look over the release-preview PR and approve a deployment. No npm keys on anyone's laptop, no checklist, no tribal knowledge.

Reviewer: ```suggestion
The biggest win of getting off the train is that releasing stops being specialized knowledge (despite efforts to write it all down). With [`release-plan`][release-plan], _any_ maintainer can trivially cut a release: they look over the release-preview PR and approve a deployment.
```

### Extra release branches are overhead

Keeping `beta`, `release`, and the LTS branches alive next to the default branch means constant backporting and branch bookkeeping that only exists to service the shape of the train. A single `main` gets rid of all of it.

### release-plan automates the mechanical work

Working out the next version number and assembling the changelog are exactly the mechanical, error-prone steps that shouldn't be done by hand, and this is the important bit: release-plan does this for us, deterministically, from the labels and titles of the PRs that got merged. The ecosystem has already standardized on release-plan for almost every addon, so the framework doesn't need a bespoke process that's strictly more work than the tooling everyone else already trusts.

### Keep the cadence

The six-week cadence is predictable, well understood, and well loved. It's _not_ what makes the train expensive, so we deliberately keep it. We're moving off the branch-and-process model, not the calendar.

## Detailed design

### Branching model

> One long-lived branch: `main`. Every PR merges here, so `main` is both where work integrates and the source of every release. There are no `beta` or `release` branches, and none of the backporting that keeping multiple release branches requires.

Reviewer: main is always the long-lived branch, don't use redundant phrasing

### Per-PR release metadata

> release-plan is label-driven. That's the one way it works, so there's nothing to invent here. Every PR gets one of its labels:

Reviewer: just lay release plan uses these labels:

- LTS keeps working like it does today.

So this RFC changes _how_ a release is cut and _which branches exist_, not _when_ releases happen or _what_ they guarantee.

### Lockstep across packages

`ember-source` and `ember-cli` keep releasing in lockstep, but lockstep here is a timing thing, not a coupling. Each repo releases independently through release-plan (its natural per-repo mode), and the releases don't need to know about each other. Since both cut stable on the same six-week date, their versions stay lined up. There's no cross-repo coordination step in the workflow; the shared cadence is what keeps them in step.

### What this removes

- The `beta` and `release` branches, and the `ember-source@beta` dist-tag.
- The backporting needed to keep multiple release branches alive.
- The bespoke per-cycle release process. Version bumps, changelog assembly, and the cut/promote steps are all release-plan's job now.

> ### What this keeps

Reviewer: also add consumers will not notice a difference

> ### What this keeps

Reviewer: keeps: backports to LTSes, when needed

- The six-week release cadence, unchanged.
- SemVer and every compatibility guarantee we make today.
- The deprecation policy, LTS, and the major-version process (RFC #0830).

> - Steering's control over majors and over what ships.

Reviewer: steering doesn't decide this. remove all mention of "Steering"

`beta` users are the folks whose workflow moves. If your `ember-try` config or CI matrix tests `ember-source@beta`, you remove that scenario, since `@beta` won't exist anymore. You can add `@alpha` if you want a published prerelease. Canary testing (consuming `main` from git) is unaffected.

Docs work:

- Rewrite the website Releases page: `main` is the release line, the six-week cadence is unchanged, the prerelease stream is `@alpha` (published nightly from `main`), and `beta` is gone.
- Update the contributor guide with the labeling workflow.
- Document the environment, who can approve, and how approval works.
- A migration / announcement blog post about dropping `@beta`.

## Drawbacks

- Dropping `@beta` moves some consumers. Any `ember-try` scenario or addon CI matrix pinned to `ember-source@beta` has to remove that scenario (and can adopt `@alpha` if it wants). It's a one-time, mechanical cleanup rather than a lost capability, since canary testing is unaffected, but it's still ecosystem-wide churn that needs coordinating.

> - Collapsing `beta` into `alpha` loses a soak stage. Today `beta` is a distinct checkpoint between the default branch and stable. With one prerelease stream instead of two, changes get less differentiated baking before a scheduled stable release.

Reviewer: remove this line

> - Labeling discipline matters. A wrong label means a wrong bump. release-plan makes the bump deterministic, but the label itself is human-supplied.

Reviewer: remove this line

> - Publish authority. The folks who can approve the protected environment are the same active people who cut releases today (informal, no change), but it does mean the npm publish is only as locked down as that environment's rules and those accounts' security (2FA / OIDC).

Reviewer: not an issue, remove this

> - Tooling dependency. The framework's release process becomes coupled to release-plan. It's a small, community-owned tool, but it's a new dependency.

Reviewer: we kind of also own this, so it's not a big deal

---
stage: accepted
start-date: 2026-06-05T00:00:00.000Z
release-date: # In format YYYY-MM-DDT00:00:00.000Z
release-versions:
teams: # delete teams that aren't relevant
  - cli
  - framework
  - steering
prs:
  accepted: # Fill this in with the URL for the Proposal RFC PR
project-link:
suite:
---

# Move off the release train to a simpler release model

## Summary

Ember's "release train" today bundles two separate things:

1. **A six-week release cadence.** A new minor ships on a steady, predictable
   clock.
2. **A multi-channel promotion pipeline.** Code flows through dedicated
   `canary` → `beta` → `release` branches, cut and promoted by a rotating
   release manager working a mostly-manual checklist each cycle.

**This RFC keeps (1) and replaces (2).** The six-week cadence is the part that
works, and it stays exactly as it is. What goes away is the multi-channel branch
machinery and the hand-run release ceremony.

Concretely:

> - A single long-lived branch: **`main`**. No `canary`, `beta`, or `release`

Reviewer: there's never been a canary branch

  branches.
- Stable releases are cut from `main` by [`release-plan`][release-plan] on the
  same six-week schedule, with the `npm publish` gated behind a protected
  [GitHub deployment environment][gh-environments] — a required reviewer
  approves before it publishes.
- The **`canary` and `beta` channels collapse into a single `alpha`**, published

>   automatically from `main` (the way Embroider published its prereleases).

Reviewer: the way embroider and glint _used to_ publish their pre-releases

>   Because `ember-source` lives in git, `main` is also directly consumable as a

Reviewer: remove this statement

  git ref. Either way, "bleeding edge" is just "whatever is on `main`" — there is
  no separate channel branch or nightly promotion job behind it.

SemVer, the six-week cadence, the deprecation policy, LTS, and the major-version
process from RFC [#0830][rfc-830] are all **unchanged**. Only the branch
structure and the publishing mechanics change.

[release-plan]: https://github.com/embroider-build/release-plan
[gh-environments]: https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment
[rfc-830]: https://github.com/emberjs/rfcs/blob/master/text/0830-evolving-embers-major-version-process.md

## Motivation

> The expensive part of the release train is not the cadence — it's the machinery

Reviewer: and the time it takes maintainers with limited time to manage it all

around it.

### Three live channels are a standing branch-management cost

Maintaining `canary`, `beta`, and `release` as separate, simultaneously-live
lines (plus LTS) means ongoing cherry-picks, back-merges, and branch bookkeeping

> that exist only to service the shape of the train. Every cycle a release manager

Reviewer: we don't have a formal release manager, it's really just one person -- katie for ember.js and chris for ember-cli

getting rid of the release-trains would mean that _any maintainer_ can trivially do releases

(RM) works a long, mostly-manual checklist: cut `beta` from `canary`, promote

> `beta` to `release`, hand-publish `ember-source`, `ember-cli`, and (historically)

Reviewer: don't use the phrase hand-publish, we publish from CI

> `ember-data` in lockstep, regenerate and proofread the changelog, and coordinate

Reviewer: don't mention ember-data

the release. That work:

- **Needs a volunteer on a clock, indefinitely**, and slips when no RM is
  available. It is a recurring source of burnout and bus-factor risk.

> - **Is manual and therefore error-prone** — version bumps, changelog assembly,

Reviewer: this is a very important point that we should emphacise, and mention that release-plan handles this for us

  and multi-package lockstep publishing are exactly the mechanical steps tooling
  does more reliably than a person at the end of a checklist.

### The ecosystem already automated this

Nearly every modern Ember addon releases with [`release-plan`][release-plan]. It
derives the SemVer bump and changelog from merged PRs and publishes to npm with
provenance, no scheduled human ceremony required. The framework does not need a
bespoke, hand-run process that is strictly more work than the tooling everyone
else already trusts. **We can get away with `release-plan`.**

> ### A separate `canary` branch mostly duplicates "the git repo"

Reviewer: no, remove this section

The reason `canary` exists is so people can run unreleased Ember — and it is
already consumed straight from git, not from an `ember-source@canary` npm
dist-tag. But "the latest unreleased code" is exactly what the default branch
*is*. A separate `canary` branch plus its nightly promotion job is extra
plumbing to approximate "whatever is on `main` right now." Publishing an `alpha`
prerelease automatically from `main` (and keeping the existing git-ref workflow)
delivers the same capability with none of the branch machinery.

> ### Keep the cadence — it works

Reviewer: don't add subheadings to headings

The six-week cadence is predictable, well-understood, and well-loved. It is *not*
what makes the train expensive, so this proposal deliberately keeps it. We are
moving off the *channel-and-ceremony* model, not the calendar.

## Detailed design

### Branching model

A single long-lived branch: **`main`**. All pull requests merge here. `main` is
both where work integrates and the source of every release. There are no

> `canary`, `beta`, or `release` branches, and none of the cherry-pick /

Reviewer: there never was a canary branch

back-merge machinery that keeping three live channels requires.

### Per-PR release metadata

Every PR that changes shipped code declares:

> 1. **A SemVer impact** — `patch`, `minor`, or `breaking` (major) — via a label

Reviewer: read up on release-plan, we only have one way to do this, and it's label-driven, don't add fluff

   (or a changeset file, whichever the team prefers; `release-plan` supports the
   label-driven flow out of the box).
2. **A changelog entry** — the human-readable "what changed," authored as part
   of review rather than reconstructed afterward.

`release-plan` aggregates these across everything merged since the last release
to compute the next version number and assemble the changelog. The decision a
release manager used to make by reading the diff is now made, deterministically,
from metadata that was reviewed when the change landed.

> Breaking changes (`breaking` label) still require the same approval they require

Reviewer: this is not manual, we still label a PR on the 12 X 6 week cycle as our normal major release cadence enables
